File: fastlogging/examples/spawn.rs
use std::{
    process::{self, Command},
    thread,
    time::Duration,
};

use fastlogging::{root, LoggingError};

fn run_parent(child: u32) -> Result<(), LoggingError> {
    println!("# {} Run parent. Child has pid {child}.", process::id());
    root::debug("Debug Message from parent")?;
    root::info("Info Message from parent")?;
    root::warning("Warning Message from parent")?;
    root::error("Error Message from parent")?;
    root::fatal("Fatal Message from parent")?;
    println!("# {} Parent finished", process::id());
    Ok(())
}

fn run_child(ppid: u32) -> Result<(), LoggingError> {
    println!("# {} Run child. Parent has pid {ppid}.", process::id());
    thread::sleep(Duration::from_millis(20));
    root::debug("Debug Message from child")?;
    root::info("Info Message from child")?;
    root::warning("Warning Message from child")?;
    root::error("Error Message from child")?;
    root::fatal("Fatal Message from child")?;
    println!("# {} Child finished", process::id());
    Ok(())
}

fn main() -> Result<(), LoggingError> {
    println!("\n# {} Start main", process::id());
    //fastlogging::set_debug(3);
    if let Some(ppid) = root::get_parent_pid() {
        // This is the child
        run_child(ppid)?;
    } else {
        // This is the parent. Create child process.
        let mut child = Command::new(std::env::current_exe()?)
            .spawn()
            .expect("failed to execute child");
        run_parent(child.id())?;
        child.wait()?;
    }
    println!("# {} Continue main", process::id());
    root::debug("Debug Message from main")?;
    thread::sleep(Duration::from_millis(10));
    println!("# {} SyncAll main", process::id());
    root::sync_all(0.1)?;
    println!("# {} Shutdown main", process::id());
    root::shutdown(false)?;
    println!("# {} Finished main\n", process::id());
    Ok(())
}

File: pyfastlogging/examples/spawn.py
import os
import time

import pyfastlogging as fl


def run_parent():
    print("# Run parent.")
    fl.debug("Debug Message from parent")
    fl.info("Info Message from parent")
    fl.warning("Warning Message from parent")
    fl.error("Error Message from parent")
    fl.fatal("Fatal Message from parent")
    print("# Parent finished")


def run_child(ppid: int):
    print(f"# Run child with pid {os.getpid()}. Parent has pid {ppid}.")
    time.sleep(0.02)
    fl.debug("Debug Message from child")
    fl.info("Info Message from child")
    fl.warning("Warning Message from child")
    fl.error("Error Message from child")
    fl.fatal("Fatal Message from child")
    print("# Child finished")


if __name__ == "__main__":
    import multiprocessing

    multiprocessing.set_start_method("spawn")
    print(f"# Start main with pid {os.getpid()}")
    if ppid := os.fork():
        run_child(ppid)
    else:
        run_parent()
    print(f"# Continue main with pid {os.getpid()}")
    fl.debug("Debug Message from main")
    print("# main finished")
    time.sleep(0.1)
